Widgets: Share store TZ settings

ealeksandrov opened the pull request:
Closes: #7564

# Description
To make sure we're using store TZ, share it as a parameter to widget extension and use in stats requests.

# How

- Adds `defaultStoreTimeZoneGMTOffset` userdefault property
- Moves `Date+StartAndEnd` extension to shared WooFoundation module
- Uses shared extension with store timezone in API requests

*Note:* Testing this is complicated and sometimes I got different stats result in app dashboard and widget.
Stats dashboard uses store TZ for UI, but device TZ for requests:
https://github.com/woocommerce/woocommerce-ios/blob/29c95b9bc7e1424ca4e0f59eb9eb5649848dfe3f/WooCommerce/Classes/ViewRelated/Dashboard/Stats%20v4/StoreStatsAndTopPerformersViewController.swift#L206-L210

Should we do the same?

# Testing Steps

- Set timezone on device and store to be very different. In `wp-admin` it's in Settings -> General -> Timezone.
- Add orders close to days transitions in both TZs.
- Launch the app.
- Compare stats data on app dashboard and widget.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ealeksandrov committed:
Share store TZ to extension

ealeksandrov committed:
Move Date extension to WooFoundation

ealeksandrov committed:
Move extension test

ealeksandrov committed:
Use site TZ in StoreInfoDataService

ealeksandrov committed:
Merge branch 'trunk' into issue/7564-store-settings

ealeksandrov commented on WooCommerce/Classes/Yosemite/DefaultStoresManager.swift:
```diff
@@ -510,6 +510,7 @@ private extension DefaultStoresManager {
         // Non-critical store info
         UserDefaults.group?[.defaultStoreID] = siteID
         UserDefaults.group?[.defaultStoreName] = sessionManager.defaultSite?.name
+        UserDefaults.group?[.defaultStoreTimeZoneGMTOffset] = sessionManager.defaultSite?.siteTimezone.secondsFromGMT() ?? TimeZone.current.secondsFromGMT()
```
This presentation is picked because we init it same way from API response:
https://github.com/woocommerce/woocommerce-ios/blob/29c95b9bc7e1424ca4e0f59eb9eb5649848dfe3f/Networking/Networking/Model/Site.swift#L168-L173

Ecarrion commented on WooCommerce/StoreWidgets/StoreInfoDataService.swift:
```diff
@@ -25,10 +25,15 @@ final class StoreInfoDataService {
     private var network: AlamofireNetwork
 
-    init(authToken: String) {
+    /// Timezone of the website
```
```suggestion
    /// Timezone of the store
```

Ecarrion reviewed (commented):
Code looks good!

> Should we do the same?

I think we should match the way the app fetches & displays information. Because if not the merchant won't know what source of information to trust.

App | Widget
--- | ---
![IMG_4700](https://user-images.githubusercontent.com/562080/189758608-18309c2e-e54d-4d84-a7f8-f026a498bc9f.PNG) | ![IMG_4701](https://user-images.githubusercontent.com/562080/189758582-72338e46-75bf-4d43-8142-dca8f44c089f.PNG)

> Since the stats charts are based on site time zone, we set the time zone for the stats UI to be the site time zone. 

 Wouldn't that means that we don't need the site timezone? As we don't show charts?

ealeksandrov commented:
> Wouldn't that means that we don't need the site timezone? As we don't show charts?

That's what I think now. Should we close it and maybe get back if we found inconsistencies in testing?

Update: Actually, I think `Date+StartAndEnd` reuse worth merging. So I reverted TZ sharing (49687de) and updated code to use device TZ (eedc08e).

ealeksandrov committed:
Revert "Share store TZ to extension"

76eac481bb8d3c59558998df8d89447919e318e3

ealeksandrov committed:
Replace store TZ with device TZ in widgets API requests

Ecarrion commented on WooCommerce/StoreWidgets/StoreInfoDataService.swift:
```diff
@@ -74,41 +73,18 @@ private extension StoreInfoDataService {
                 siteVisitStatsRemote.loadSiteVisitorStats(for: storeID,
                                                           unit: .day,
-                                                          latestDateToInclude: Date.startOfToday,
+                                                          latestDateToInclude: Date().endOfDay(timezone: .current),
```
Wow I don't know why I used `startOfToday` 🤦

Ecarrion reviewed (approved):
> Update: Actually, I think Date+StartAndEnd reuse worth merging. So I reverted TZ sharing (49687de) and updated code to use device TZ (eedc08e).

Agree specially since we recently added some logs due to a new crash around date parsing.
